src/ol_infrastructure/components/services/vault.py

```diff
@@ -134,3 +137,81 @@ def __init__(
                 default_ttl=db_config.default_ttl
             )
         self.register_outputs({})
+
+class OLVaultAWSSecretsEngineConfig(BaseModel):
+    app_name: Text
+    access_key: Text
+    default_lease_ttl_seconds: int = SIX_MONTHS
+    max_lease_ttl_seconds: int = SIX_MONTHS
+    description: Text
+    secret_key: Text
+    path: Text
+    policy_documents: Dict
+    credential_type: str = 'iam_user'
+
+
+class OLVaultAWSSecretsEngine(ComponentResource):
+
+    def __init__(
+            self,
+            engine_config: OLVaultAWSSecretsEngineConfig,
+            opts: ResourceOptions = None
+    ):
+        super().__init__('ol:services:VaultAWSSecretsEngine', engine_config.app_name, None, opts)
+
+        self.aws_secrets_engine = aws.SecretBackend(
+            # TODO verify app_name exists based on Apps class in ol_types
+            f'aws-{engine_config.app_name}',
+            access_key=engine_config.access_key,
+            secret_key=engine_config.secret_key,
+            path=f'aws-{engine_config.app_name}'
+        )
+
+        for role_name, policy in engine_config.policy_documents.items():
+            aws.SecretBackendRole(
+                role_name,
+                backend=f'aws-{engine_config.app_name}',
+                credential_type='iam_user',
+                name=role_name,
+                policy_document=json.dumps(policy),
+                opts=ResourceOptions(depends_on=[self.aws_secrets_engine])
+            )
+
+        self.register_outputs({})
+
+
+class OLVaultAppRoleAuthBackendConfig(BaseModel):
+    name: Text
+    type: str = 'approle'
+    description: Text
+    backend: Text
+    role_name: Text
+    token_policies: List
+
+
+class OLVaultAppRoleAuthBackend(ComponentResource):
+
+    def __init__(
+            self,
+            approle_config: OLVaultAppRoleAuthBackendConfig,
+            opts: ResourceOptions = None
+    ):
+        super().__init__('ol:services:VaultAppRoleAuthBackend', approle_config.name, None, opts)
+
+        self.approle_backend = AuthBackend(
+            approle_config.name,
+            description=approle_config.description,
+            path=approle_config,
+            type=approle_config.type)
+
+        for policy in approle_config.token_policies:
+            self.approle_policy = Policy(f'{approle_config.name}', policy=policy)
+
+        self.approle_backend_role = approle.AuthBackendRole(
+            approle_config.name,
+            backend=self.approle_backend.path,
+            role_name=approle_config.role_name,
+            token_policies=approle_config.token_policies,
```

Comment:
Might be good to populate a list of policies in the loop above and then pass that list here so that there is a direct dependency of the order of resource creation.